### DISCUSSION TITLE
Guide/Instructions/Keybind Observations and Suggestions

### DISCUSSION THREAD

**Discussion by Dark-Sol** [Xithrius/twitch-tui#398] [General]
Hey @Xithrius!

Thanks for hearing me out on the Python Discord.  Here are some of the things that I had to stumble through or found when setting up twitch-tui:

1. For someone who isn't familiar with Rust, I wasn't sure how to write out the channel names in the array as described in config.toml. (I was able to figure out that the channel names need to be within double quotes and separated by commas, but telling the user how to do that in the guide would help.)
2. With the above, I think there is a 5 or 6 channel limit to how many will show up in the main menu. That isn't mentioned anywhere, so I initially loaded up that array with _all_ of my channel favourites and then was wondering why only 5 would show up. That should also be mentioned in the guide.
3. When doing `twt -h`, is `-t` missing as an option? `--theme <THEME>` is there, but not the shortcut.
4. When using `-s`, the documentation doesn't explain what "start", "normal" or "help" do. I just had to try each option to see what it did. 
    - Additionally, the guide doesn't go into depth about what all of the input options are or what all the configuration items in config.toml do.
5. I think if you're going to use `twt -h` for help at the prompt, maybe consider using `h` inside the application for help as well? Having to remember that one is `h` while the other is `?` could confuse users.
6. `@` and `/` don't seem to be doing anything when I try to enter Insert Mode. (I'll add an issue for this.)
7. I would switch `i` to `c`. It feels like it could be more intutive for a user if you called Insert Mode "Chat Mode" and had the user press `c` to enter "Chat Mode".
    - `c` doesn't seem to do anything when I try using it, as an aside. (I'll add an issue for this as well.)
8. Consider changing the description for `Esc` to "Go back to previous window" for clarity?
9. Consider adding a keybind to add the current channel to the "Configured default channels".
    - This is more of a UI/UX thing, which I'll also put in my other discussion post, but I would change the heading for this section to "Favorite Twitch Streamers" or something like that.

I'm sure I can think of some other things, but this is probably a good first pass.

**Comment by Xithrius** [comment]
Thank you very much for the feedback. I'll implement these suggestions, then we can do another pass.

**Comment by Xithrius** [comment]
Sorry for the delay, but I will start actually going over the suggestions listed here. Shouldn't have much of an issue implementing most of these.

In the meantime, I noticed that for point 3, `-t` is already used. `--tick-delay` already has ownership of it, so `--theme` can't have it. One of them will have to change, and it'll probably be tick delay.

**Reply by Xithrius** [reply]
On another note, point 6 has been fixed in b9d2cdeaecbc6135f9baee11a839cac1b7f1b4c7

**Reply by Dark-Sol** [reply]
Would you consider changing `--tick-delay` to `--delay` and use `-d` for that so you can use `-t` for `--theme`?

**Reply by Xithrius** [reply]
Sounds good to me, I'll get that out today.

**Reply by Dark-Sol** [reply]
Also, no apologies necessary. This is your project and on your timeline. Other things can and will take priority. I have no reason to be impatient. 🙂

**Reply by Xithrius** [reply]
Alright, point 3 is fixed in f69ce61db177a2d82f7ed9c39259ea7776647a6e

**Reply by Xithrius** [reply]
Points so far for tracking:

1. 85ae0f065ae80c4e7d13990533c6f534d789f92f
2. f481d9e54723010998b6df0e55cc66db621d1f1c
3. f69ce61db177a2d82f7ed9c39259ea7776647a6e
4. Partially done, and I'll add screenshots with descriptions of what's going on 
5. 78dfe574f0e37179ab3215caa97e97e2042a1a4b
6. b9d2cdeaecbc6135f9baee11a839cac1b7f1b4c7
7. 319dd347e4df0d303ada921238051b3da1deb9d5
8. 319dd347e4df0d303ada921238051b3da1deb9d5
9. f481d9e54723010998b6df0e55cc66db621d1f1c

**Reply by Xithrius** [reply]
Some notes on the previous changes:

- The chat box can be pulled up with either `c` or `i`.
- The help popup can be accessed through `?` or `h`.
- Recent channels now has a user-defined restriction by `recent_channel_count` in the `[frontend]` section of the config. `start_screen_channels` has been renamed to `favorite_channels` in the config, and has no restriction on how many channels you can add. However, you can probably break the UI if you go longer than the height of the terminal can withstand.
- The config has examples, and is much more readable than before.
- `start_state` has been renamed to `first_state` to avoid collisions in short CLI args. Same reasoning for `tick_delay` to `delay`.
- `date_shown` has been renamed to `datetime_format` since that's more accurate.
- `@` and `/` have been fixed, and insert the respective character when entering the message box through said keys.
- Dashboard keybinds have been added to the help popup.

I'm going to leave setting the current channel as a favorite until later, since I'm going to try and do an overhaul of how you can choose/edit/delete them.

**Reply by Dark-Sol** [reply]
Could it be possible to put the favorite channels in multiple columns to overcome the terminal height issue? Since channel/user names can only have a max of 25 characters, you could theoretically "spill" the favorite channel list into multiple columns once the number goes over some arbitrary number. Since you're able to resize the terminal window, it only shows multiple columns if resized larger than some amount of characters. (Like collapsible/expandable columns depending on terminal width?) Or use +/- to change "pages" of a user's favorites? Like the following:

`[-] Previous 5 favorites`
`[1] Streamer_1`
`[2] Streamer_2`
`[3] Streamer_3`
`[4] Streamer_4`
`[5] Streamer_5`
`[+] Next 5 favorites`

And have the `[-]` option "grey out" and be disabled if you're at the beginning of the list and the `[+]` do the same when you're at the end?

Also, would it be possible to query Twitch for a user's followed streams that are live, since you're using OAuth already, and put those into the favorite channels list if a user doesn't want to manually add them in themselves?

Would you consider changing `datetime_format`to `timestamp_format`, since that describes the function even more clearly?

**Reply by Xithrius** [reply]
The OAuth issue has been up for a while (#147). I might try redoing the authentication system by making a popup in the browser, we'll have to see. Last time I looked at the API, authentication and fetching followers took a bunch of steps, and also required more authentication besides the token. I'll take another look, to see what comes of it.

**Reply by Xithrius** [reply]
All it took was adding another scope to the token to get the followers for the logged in user. I finally know how this darn Twitch API works.
